repose-aggregator/docs/src/asciibinder/services/http-logging.adoc

```diff
@@ -0,0 +1,359 @@
+
+| inboundRequestUrl
+| String
+| The full inbound request URL minus the query portion.
+For example, for a request with the request line `GET /path?when=now HTTP/1.1` and the header `Host: example.com`, this value would be `http://example.com/path`.
+
+| outboundRequestUrl
+| String
+| The full outbound request URL minus the query portion.
+For example, for a request with the request line `GET /path?when=now HTTP/1.1` and the header `Host: example.com`, this value would be `http://example.com/path`.
+
+| inboundRequestHeaders
+| Map<String, List<String>>
+a| The headers on the inbound request.
+The keys in this map are the header names while the values are lists of corresponding header values.
+The keys (i.e., header names) are case-sensitive and will only contain lowercase letters.
+Headers with multiple values spanning multiple lines will have a value list with one item for each line.
+For example, given the header lines:
+```
+Test-Header: foo, bar
+Test-Header: baz
+```
+The value of this variable would be:
+```
+{ "test-header": ["foo, bar", "baz"] }
+```
+
+| outboundRequestHeaders
+| Map<String, List<String>>
+a| The headers on the outbound request.
+The keys in this map are the header names while the values are lists of corresponding header values.
+The keys (i.e., header names) are case-sensitive and will only contain lowercase letters.
+Headers with multiple values spanning multiple lines will have a value list with one item for each line.
+For example, given the header lines:
+```
+Test-Header: foo, bar
+Test-Header: baz
+```
+The value of this variable would be:
+```
+{ "test-header": ["foo, bar", "baz"] }
+```
+
+| outboundResponseHeaders
+| Map<String, List<String>>
+a| The headers on the outbound response.
+The keys in this map are the header names while the values are lists of corresponding header values.
+The keys (i.e., header names) are case-sensitive and will only contain lowercase letters.
+Headers with multiple values spanning multiple lines will have a value list with one item for each line.
+For example, given the header lines:
+```
+Test-Header: foo, bar
+Test-Header: baz
+```
+The value of this variable would be:
+```
+{ "test-header": ["foo, bar", "baz"] }
+```
+
+| outboundResponseStatusCode
+| Integer
+| The status code of the outbound response.
+For example, for a response with the status line `HTTP/1.1 200 OK`, this value would be `200`.
+
+| outboundResponseReasonPhrase
+| String
+| The reason phrase of the outbound response.
+For example, for a response with the status line `HTTP/1.1 200 OK`, this value would be `OK`.
+
+| outboundResponseContentLength
+| Integer
+| The size of the outbound response body in bytes.
+This value is taken from the `Content-Length` header of the response.
+For example, for a response with the header line `Content-Length: 1234`, this value would be `1234`.
+
+| outboundResponseBytesWritten
+| Integer
+| The number of bytes of the outbound response body written.
+
+| userId
```

Comment:
These values shouldn't be from the outbound request. They should be written directly by the keystone filter. In fact this implies if the keystone filter rejects the request for authorization reasons these wouldn't get populated, which is super not the behavior we want.

Reply:
While I agree in principle that we want to capture Identity data as soon as possible, I have a number of concerns about the implementation as you have described it:

- To my understanding, it has been the plan to moved authorization out of the Keystone v2 filter. In the case, the reasons for a request being rejected from the Keystone v2 filter should be limited to a failure to retrieve data, in which case we don't have data to populate anyway. To split hairs, it's true that we could succeed on the validation call and fail on a different call, though I believe we actually handle that right now not by rejecting the request, but by not populating the data we could not acquire. In other words, if the Keystone v2 filter will only reject a request if it cannot retrieve user data from Identity, then we don't have to worry about this issue, and that may be the case once we remove authorization from that filter.
- The difference in granularity is minimal (in-filter versus after-filter). Since we have a general way of capturing data after-filter, I certainly prefer relying on that over a filter-specific solution. Doing so reduces the amount of code, the code complexity, and the duplication of data.
- We can always change what data we capture and where we capture it in the future. Which is to say, I believe this implementation may fall outside the scope of this story. Certainly it is not specified in the acceptance criteria.

I understand that the primary concern is the case where the Keystone v2 filter rejects a request, but keeping in mind the points above, I believe that concern is addressed by the (temporarily) continued existence of delegation and by the plan to move authorization logic. When we remove delegation, we will need to make sure we have a way of reliably capturing Identity data, but I see that as work outside the scope of this story. Personally, I am not a fan of how inflated some of our initial implementation stories get, and this is one instance where I was trying to trim the fat. And in truth, I would like more time to consider this issue so that I can hopefully devise an acceptable general solution that does not involve instrumenting every data-producing filter that may reject a request.

Reply:
If we do decide to move forward with capturing user data as described, we would likely do so here:
https://github.com/rackerlabs/repose/blob/master/repose-aggregator/components/filters/keystone-v2-filter/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala#L379

Basically, as soon as we we have validated the token, grab the logging context and populate user data.

Reply:
SO to me this auth information was always one of the key pieces for acceptance. If it didn't get down into the acceptance criteria that was an oversight, because it was in the list originally. also despite the fact that we are splitting them in the future the fact remains that most of our users are on the single filter solution right now, and that use case has to work, especially since this work is meant to be a replacement for the existing http slf4j filter. Which actually reminds me that it should be deprecated as well.

Reply:
Wait, what? I thought we wanted to keep the `Slf4jHttpLoggingFilter` to be able to capture data at certain points in the filter chain? This service only keeps the latest inbound and outbound messages, while the SLF4J filter allows inspection/logging of messages at designated points in the filter chain.

As for the auth information, to be clear, we are providing a way to capturing and expose that data as it stands. What we are talking about here is really making it easier for the user to capture that data in the case where the user is using the Keystone v2 filter to perform authorization (i.e., not forcing the user to use delegation). However, given that authorization in Keystone v2 is already deprecated, it seems odd to me to write code to support that use case. As soon as Keystone v2 authorization goes away, that code is no longer necessary. As long as delegation exists, that code is not necessary. If delegation and Keystone v2 authorization are both being removed in `10.0.0.0`, then we don't have a problem.

As a quick aside, the user name was certainly part of the acceptance criteria, and it was reasonable to add user ID, impersonator ID, and impersonator name.

Reply:
SO yes, its usage has been deprecated, but while it hasn't been removed we still have to support the use case. This work is on master and will roll out on repose 9.x. I really don't expect anyone to move off of it till we force the issue in 10. Are we keeping the other filter? What purpose does that serve? The only real use case we know of it has been completely superseded by this. While there is a theoretical use for it, i don't see it as one people will actually use.

I totally get that it's unfortunate to add code we know will go away soon, but there's only so much change we want to push on the users at once.

Reply:
I guess in my view we are adding a feature with this service, and a new feature does not necessarily have to accommodate a deprecated feature. Keystone v2 filter authorization is a deprecated feature, as is delegation, and in conjunction they will continue to work until they are removed. Additionally, delegation allows the new HTTP Logging Service to work as expected alongside Keystone v2 filter authorization. So if the user is currently configured to do audit logging (using delegation), that will continue working. If they want to use the new service, that will also work (allowing them to drop the HERP filter, though the DeRP filter would still be necessary). If they wanted to drop delegation and continue using Keystone v2 filter authorization that is where trouble arises, and that is the use case we would not support (where the user is trying to move off of some deprecated functionality but not other deprecated functionality). Of course, at that point, the user is also updating their configuration, so it would likely be just as easy to drop all deprecated functionality (i.e., Keystone v2 filter authorization and delegation) and move to current functionality (i,.e., Keystone v2 Authorization filter and HTTP Logging Service).

So to my mind, this is less forcing a change on the user, and more incentivizing the user to stop relying on deprecated features to make using new features easier.

At any rate, I've finished implementing what you've asked for and will push it shortly.

Reply:
So the missing FIT case is using the Keystone v2 filter without delegation where authentication succeeds and authorization fails, then the `user`/`impersonator` `id`/`name` should all still be logged?

Reply:
@wdschei That sounds correct to me.